proposals/NNNN-rangeset-and-collection-operations.md

```diff
@@ -0,0 +1,616 @@
+
+## Proposed solution
+
+This proposal adds a `RangeSet` type for representing multiple, noncontiguous ranges, as well as a variety of collection operations for creating and working with range sets.
+
+```swift
+var numbers = Array(1...15)
+
+// Find the indices of all the even numbers
+let indicesOfEvens = numbers.indices(where: { $0.isMultiple(of: 2) })
+
+// Perform an operation with just the even numbers
+let sumOfEvens = numbers[indicesOfEvens].reduce(0, +)
+// sumOfEvens == 56
+
+// You can gather the even numbers at the beginning
+let rangeOfEvens = numbers.gather(indicesOfEvens, justBefore: numbers.startIndex)
+// numbers == [2, 4, 6, 8, 10, 12, 14, 1, 3, 5, 7, 9, 11, 13, 15]
+// numbers[rangeOfEvens] == [2, 4, 6, 8, 10, 12, 14]
+
+// Reset `numbers`
+numbers = Array(1...15)
+
+// You can also build range sets by hand using array literals...
+let notTheMiddle: RangeSet = [0..<5, 10..<15]
+print(Array(numbers[notTheMiddle]))
+// Prints [1, 2, 3, 4, 5, 11, 12, 13, 14, 15]
+
+// ...or by using set operations
+let smallEvens = indicesOfEvens.intersection(
+    numbers.indices(where: { $0 < 10 }))
+print(Array(numbers[smallEvens]))
+// Prints [2, 4, 6, 8]
+```
+
+These are just a few examples; the complete proposal includes operations like inverting a range set and inserting and removing ranges, which are covered in the next section.
+
+## Detailed design
+
+`RangeSet` is generic over any `Comparable` type, and supports fast containment checks for ranges and individual values, as well as adding and removing ranges of that type. 
+
+```swift
+/// A set of ranges of any comparable value.
+public struct RangeSet<Bound: Comparable> {
+    /// Creates an empty range set.
+    public init() {}
+
+    /// Creates a range set containing the given range.
+    public init(_ range: Range<Bound>)
+    
+    /// Creates a range set containing the given ranges.
+    public init<S: Sequence>(_ ranges: S) where S.Element == Range<Bound>
+    
+    /// A Boolean value indicating whether the range set is empty.
+    public var isEmpty: Bool { get }
+    
+    /// Returns a Boolean value indicating whether the given value is
+    /// contained by the ranges in the range set.
+    ///
+    /// - Complexity: O(log *n*), where *n* is the number of ranges in the
+    ///   range set.
+    public func contains(_ value: Bound) -> Bool
+        
+    /// Returns a Boolean value indicating whether the given range is
+    /// contained by the ranges in the range set.
+    ///
+    /// This method returns `true` if and only if one of the ranges that 
+    /// comprises the range set fully contains `range`. This example shows
+    /// ranges that are and are not contained by a range set:
+    /// 
+    ///     let set: RangeSet = [0..<5, 10..<15]
+    ///     print(set.contains(1..<4))
+    ///     // Prints "true"
+    ///     print(set.contains(4..<7))
+    ///     // Prints "false"
+    ///     print(set.contains(4..<12))
+    ///     // Prints "false"
```

Comment:
IMO this set of examples is unclear, in part because of the word "fully" above.  It leaves open what the answer would be for `set.contains(1..<5)`. I think you should at least drop "fully", because in no sense does `1..<3` "contain" `2..<4`.